### PR TITLE
[SPARK-14279] Spark Version will be picked from pom.version

### PR DESCRIPTION
**Pull request by redsanket**
## What changes were proposed in this pull request?

This PR proposes to pick the version information for SPARK from pom.version.
It also proposes to include other relevant build details.
## How was this patch tested?

spark-submit --version should give the version information and also build details

### CODE REVIEW

**Comment by srowen** [comment]
This looks like a ton of complexity, with custom plugins and generated files. What problem is this solving in comparison? The JIRA doesn't say why.

**Comment by vanzin** [comment]
The custom plugin definitely looks like overkill. Can't you do something close to this using the `antrun` plugin and the `echo` task?

Also, you'd have to change the sbt build too.

**Comment by JoshRosen** [comment]
Woah! This is _way_ too complex in its current form.

**Comment by tgravescs** [comment]
the code was mostly borrowed from other apache projects (hadoop/storm).  There is a fair amount of code here but its not really that complex. one maven plugin, one version.java file and one properties file.

It solves the problem of knowing the exact version and build information without manually editing files. I want to be able to have a automated build to build spark, update the version if required and when that gets deployed somewhere i want to know exactly what that was built from.  This is basic automated CI/CD stuff with tracability.

We can definitely do this other ways so please propose alternatives.

**Comment by vanzin** [comment]
@tgravescs see my comment; using antrun + echo (or replace) would I think cover the most important parts here. No need for a custom plugin.

**Comment by tgravescs** [comment]
@vanzin can you clarify, you mean antrun + echo to generate the properties file part or more?    You still have to call the other commands to get the git info, md5 info, etc.  

Personally I like the plugin better as its more extensible and easier to do things cleanly in java then in shell. There is logic in there for different version control systems, none, svn, git.  We probably don't need svn at this point. Personally I would rather see this then writing shell or ant scripts.  Using java keeps it less os dependent as well.  Since the code is used by other projects we know it works and it wasn't very hard to copy it and slightly modify for Sparks use. 

I forgot we were still using sbt for development.  so we will have to look into that.  Last I knew sbt couldn't use the maven plugin so it will have to be handled separately.  I guess the question there is whether it needs to be real information since its not the official build tool

**Comment by vanzin** [comment]
I mean antrun to generate the properties file. I understand that the plugin can do more and is written in Java, but that argument would be a lot stronger if this were actually a published plugin instead of a piece of code that is forked throughout projects...

On the other hand, the antrun-based approach is just a few lines of code in a pom file.

**Comment by redsanket** [comment]
@vanzin The idea was to make it pluggable to be used by other projects if neccessary. If we do not want to have additional information regarding the build it can be simplified but I think it would be nice to have this information. I will look into antrun plugin thank you

**Comment by tgravescs** [comment]
So to me here the difference is shell scripts vs Java code.  Unless you know of an easier way with antrun to do the md5sum and get the git information?  All I really saw was the exec option to call out to shell.  

I would rather maintain java code then shell scripts.  Generally the java code should be easier to maintain and less brittle.

I don't see why having our own maven plugin for this is such a big deal, it means we can do whatever we want with it but at the same times other apache projects basically use the same plugin and its known to work and shouldn't really change unless we decide to add something to it.  Why is a maven plugin any different for core code that we write from scratch?  maven provide utilities for writing your own plugins, I don't see that breaking.  We can remove the svn stuff from it and it will shrink the actual code size. 

We do have to handle it in sbt, but as I said above personally I think we could simply echo out a properties file with just the version and leave the other fields empty since its just for developer builds.

@JoshRosen you made a comment about this being to complex, what were you thinking or proposing as alternative?   And why is this to complex?

Perhaps the difference here is the jira should have been more detailed to list out all of the things we wanted to get here rather then just the version.  I'd be alright with removing the md5sum if we don't want that, but I think the rest of the information is really useful for tracability.  It makes it easy to look at a jar running on a cluster and see who built it, from what repo and commit, and what is the specific version. 

Also note that this pr has 25 files changed but the majority of them are really small just switching to use VersionInfo.

**Comment by vanzin** [comment]
> Perhaps the difference here is the jira should have been more detailed to list out all of the things we wanted to get here rather then just the version. 

Yes, that would help. Because right now the plugin is basically a (very complicated) shell script written in java; the exceptions being the date (easy to get from shell and, personally, I don't even find that very interesting) and the md5 sum (which I'm not sure what you need for - and shouldn't the git hash give you pretty much the same or better info?).

You could even write an *actual* shell script and just run it from maven / sbt and it would be easier to read and much smaller than the java plugin. (Windows compatibility would be a little sketchy, though.)

**Comment by tgravescs** [comment]
Ok, Sanket will try using the antrun method.  If it can be reused between mvn and sbt that would be easier from that perspective as well.

**Comment by srowen** [comment]
What's the motivation for this change BTW? is it strictly to reduce duplication?

**Comment by tgravescs** [comment]
Sorry if that wasn't clear, the motivation is to more easily tie this to automated continuous integration and deployment and to easily have traceability.  

Part of this is right now you have to manually change a java file to change the version that comes out when you run spark-submit --version.   With continuous integration the build numbers could be something like 1.6.1.X (where X increments on each change) and I want to see the exact version easily.  Having to manually change a java file makes that hard.  obviously that should make the apache spark releases easier as you don't have to manually change this file as well.  

The other important part for me is the git information.  This easily lets me trace it back to exact commits.   We have a multi-tenant YARN cluster and users can run many different versions at once.  I want to be able to see exactly which version they are running.  The reason to know exact version can range from helping debug some problem to making sure someone didn't hack something in Spark to cause bad things (generally they should use approved version), etc.

File: core/src/main/java/org/apache/spark/VersionInfo.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+public class VersionInfo {
+
+  private static final Logger LOG = LoggerFactory.getLogger(VersionInfo.class);
```
Comment (jodersky) [inline core/src/main/java/org/apache/spark/VersionInfo.java]: this field is never used

File: core/src/main/java/org/apache/spark/VersionInfo.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+  public static String getSrcChecksum() {
+    return COMMON_VERSION_INFO._getSrcChecksum();
+  }
+
+  public static String getBuildVersion(){
+    return COMMON_VERSION_INFO._getBuildVersion();
+  }
+
+  public static void main(String[] args) {
```
Comment (jodersky) [inline core/src/main/java/org/apache/spark/VersionInfo.java]: Is the main method in this utility class ever used?

File: core/src/main/java/org/apache/spark/VersionInfo.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  protected String _getBuildVersion(){
+    return getVersion() +
+            " from " + _getRevision() +
+            " by " + _getUser() +
+            " source checksum " + _getSrcChecksum();
+  }
+
+  private static VersionInfo COMMON_VERSION_INFO = new VersionInfo("spark");
```
Comment (jodersky) [inline core/src/main/java/org/apache/spark/VersionInfo.java]: This whole class seems pretty complex. Why are the accessors protected but exposed through static methods? Would it not be simpler to just have a static getter to a single instance and make all member accessors public? In that case I would also recommend looking into implementing this as a class/singleton object.

**Comment by vanzin** [comment]
Can we close this PR since it's been superseded by #13061?

**Pull request closed without merging**